libc-bottom-half/cloudlibc/src/libc/sys/socket/recv.c

```diff
@@ -5,9 +5,185 @@
 #include <sys/socket.h>
 
 #include <assert.h>
-#include <wasi/api.h>
 #include <errno.h>
 #include <stdint.h>
+#include <wasi/api.h>
+
+#ifdef __wasilibc_use_preview2
+#include <descriptor_table.h>
+#include "__utils.h"
+
+ssize_t tcp_recvfrom(tcp_socket_t* socket, uint8_t* buffer, size_t length, int flags,
+    struct sockaddr* addr , socklen_t* addrlen)
+{
+    // TODO wasi-sockets: flags:
+    // - MSG_WAITALL: we can probably support these relatively easy.
+    // - MSG_OOB: could be shimmed by always responding that no OOB data is available.
+    // - MSG_PEEK: could be shimmed by performing the receive into a local socket-specific buffer. And on subsequent receives first check that buffer.
```

Comment:
This implementation of the [WASI-Sockets proposal](https://github.com/WebAssembly/wasi-sockets/) isn't really the place to discuss the proposal's design itself; I'd suggest you file an issue on the proposal repo instead.

That being said, the proposal has reached phase 3 in the WASI standardization process, while the concern you're bringing up here is one that really should've been brought up earlier, because it affects the proposal in fundamental ways. As such, it might be too late to change course at this point.

Reply:
well, is the comment "MSG_PEEK: could be shimmed by performing the receive into a local socket-specific buffer. And on subsequent receives first check that buffer" a part of the proposal?
i don't know if it's the only way to implement MSG_PEEK with the proposal.

Reply:
Anything related to the implementation strategy here makes sense to discuss here, of course. I was referring to your comment about the API itself.

Reply:
sure, i agree this is not the correct place to suggest changes to the api.
if the commented strategy is the only possible implementation with the proposal, i suspect some changes are necessary. but i'm not sure if it's the case or not.

Reply:
> well, is the comment (...) a part of the proposal?

No. The only thing part of the proposal is that [MSG_PEEK isn't directly supported](https://github.com/WebAssembly/wasi-sockets/blob/main/Posix-compatibility.md#peeking-msg_peek-)

`wasi:sockets` defers to `wasi:io` for its I/O (at least for TCP) and will switch over to component-model native `streams` for both TCP & UDP when the time comes. Neither `wasi:io` nor the component-model has or is currently expected to have a "peeking" operation. (nor should they, IMO). That's why wasi:sockets doesn't have a peek operation.

> i don't know if it's the only way to implement MSG_PEEK with the proposal.

I haven't explored all possibilities (hence the TODO comment), but with the current proposal all solutions would have to be in wasi-libc in some form or another. Either that or return EOPNOTSUPP like it does in this PR.


> if the commented strategy is the only possible implementation with the proposal, i suspect some changes are necessary.

Before going down that path, could you please elaborate your original comment?:

> such a shim would break tcp flow control.

Reply:
> > well, is the comment (...) a part of the proposal?
> 
> No. The only thing part of the proposal is that [MSG_PEEK isn't directly supported](https://github.com/WebAssembly/wasi-sockets/blob/main/Posix-compatibility.md#peeking-msg_peek-)
> 
> `wasi:sockets` defers to `wasi:io` for its I/O (at least for TCP) and will switch over to component-model native `streams` for both TCP & UDP when the time comes. Neither `wasi:io` nor the component-model has or is currently expected to have a "peeking" operation. (nor should they, IMO). That's why wasi:sockets doesn't have a peek operation.

ok. thank you for explanation.
why do you think they shouldn't?

> 
> > i don't know if it's the only way to implement MSG_PEEK with the proposal.
> 
> I haven't explored all possibilities (hence the TODO comment :) ), but with the current proposal all solutions would have to be in wasi-libc in some form or another. Either that or return EOPNOTSUPP like it does in this PR.
> 
> > if the commented strategy is the only possible implementation with the proposal, i suspect some changes are necessary.
> 
> Before going down that path, could you please elaborate your original comment?:
> 
> > such a shim would break tcp flow control.

as you actually read (vs peek) from the socket, tcp can advertise the new recv window to the peer.